Core/Registry/IRegistryQuerier.cs

```diff
@@ -0,0 +1,144 @@
+        ///     If no KSP version is provided, the latest module for *any* KSP version is given.
+        /// </summary>
+        List<CkanModule> LatestAvailableWithProvides(string identifier, KSPVersion ksp_version, RelationshipDescriptor relationship_descriptor = null);
+
+        /// <summary>
+        ///     Checks the sanity of the registry, to ensure that all dependencies are met,
+        ///     and no mods conflict with each other.
+        /// <exception cref="InconsistentKraken">Thrown if a inconsistency is found</exception>
+        /// </summary>
+        void CheckSanity();
+
+        /// <summary>
+        /// Finds and returns all modules that could not exist without the listed modules installed, including themselves.
+        /// </summary>
+        HashSet<string> FindReverseDependencies(IEnumerable<string> modules);
+
+
+        /// <summary>
+        /// Gets the installed version of a mod. Does not check for provided or autodetected mods.
+        /// </summary>
+        /// <returns>The module or null if not found</returns>
+        Module GetInstalledVersion(string identifer);
+
+        /// <summary>
+        /// Attempts to find a module with the given identifier and version.
+        /// </summary>
+        /// <returns>The module if it exists, null otherwise.</returns>
+        CkanModule GetModuleByVersion(string identifier, Version version);
+
+        /// <summary>
+        ///     Returns a simple array of all incompatible modules for
+        ///     the specified version of KSP.
+        /// </summary>
+        List<CkanModule> Incompatible(KSPVersion ksp_version);
+
+        /// <summary>
+        /// Returns a dictionary of all modules installed, along with their
+        /// versions.
+        /// This includes DLLs, which will have a version type of `DllVersion`.
+        /// This includes Provides if set, which will have a version of `ProvidesVersion`.
+        /// </summary>
+        Dictionary<string, Version> Installed(bool include_provides = true);
+
+        /// <summary>
+        /// Returns the InstalledModule, or null if it is not installed.
+        /// Does *not* look up virtual modules.
+        /// </summary>
+        InstalledModule InstalledModule(string identifier);
+
+        /// <summary>
+        /// Returns the installed version of a given mod.
+        ///     If the mod was autodetected (but present), a version of type `DllVersion` is returned.
+        ///     If the mod is provided by another mod (ie, virtual) a type of ProvidesVersion is returned.
+        /// </summary>
+        /// <param name="with_provides">If set to false will not check for provided versions.</param>
+        /// <returns>The version of the mod or null if not found</returns>
+        Version InstalledVersion(string identifier, bool with_provides = true);
+    }
+
+    /// <summary>
+    /// Helpers for <see cref="IRegistryQuerier"/>
+    /// </summary>
+    public static class IRegistryQuerierHelpers
```

Comment:
You are correct on the naming of the helper class - I'm just used to
postfixing helper and forgot about the I prefix. Given that it only
contains extension methods a postfix of extensions may be better then
helper anyway. I'm currently interstate so I can't make a PR untill Sunday.
 On 26/06/2015 9:56 pm, "Paul Fenwick" <notifications@github.com> wrote:

> In Core/Registry/IRegistryQuerier.cs
> <https://github.com/KSP-CKAN/CKAN/pull/1167#discussion_r33349249>:
>
> > +        InstalledModule InstalledModule(string identifier);
> > +
> > +        /// <summary>
> > +        /// Returns the installed version of a given mod.
> > +        ///     If the mod was autodetected (but present), a version of type `DllVersion` is returned.
> > +        ///     If the mod is provided by another mod (ie, virtual) a type of ProvidesVersion is returned.
> > +        /// </summary>
> > +        /// <param name="with_provides">If set to false will not check for provided versions.</param>
> > +        /// <returns>The version of the mod or null if not found</returns>
> > +        Version InstalledVersion(string identifier, bool with_provides = true);
> > +    }
> > +
> > +    /// <summary>
> > +    /// Helpers for <see cref="IRegistryQuerier"/>
> > +    /// </summary>
> > +    public static class IRegistryQuerierHelpers
>
> I'm new to the culture, but is IRegistryQuerierHelpers the accepted way
> of doing this? The I immediately makes me think this is a virtual
> (interface) class, except that obviously that's not the case here.
>
> (In Moose we'd compile the virtual class and the helper into a *role*,
> but obviously we can't do that in C#. :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/KSP-CKAN/CKAN/pull/1167/files#r33349249>.
>